src/Pact/Typechecker.hs

```diff
@@ -736,16 +736,16 @@ trackNode ty i = trackAST node >> return node
   where node = Node i ty
 
 -- | Main type transform, expecting that vars can only refer to a user type.
-toUserType :: Term (Either Ref (AST Node)) -> TC UserType
+toUserType :: forall n . Show n => Term (Either Ref n) -> TC UserType
 toUserType t = case t of
   (TVar (Left r) _) -> derefUT r
   _ -> die (_tInfo t) $ "toUserType: expected user type: " ++ show t
   where
-    derefUT (Ref r) = toUserType' r
+    derefUT (Ref r) = toUserType' (fmap Left r :: Term (Either Ref n))
     derefUT Direct {} = die (_tInfo t) $ "toUserType: unexpected direct ref: " ++ show t
 
-toUserType' :: Show n => Term n -> TC UserType
-toUserType' TSchema {..} = Schema _tSchemaName _tModule <$> mapM (traverse toUserType') _tFields <*> pure _tInfo
+toUserType' :: Show n => Term (Either Ref n) -> TC UserType
+toUserType' TSchema {..} = Schema _tSchemaName _tModule <$> mapM (traverse toUserType) _tFields <*> pure _tInfo
```

Comment:
Is this line the crux of the change? Calling `toUserType` instead of `toUserType'`?

Reply:
We get `Term Ref`s from the loaded module, and need to convert them to `AST UserType`, using an Either along the way for things that have been already converted. `toUserType`, which calls `toUserType'`, perform the coercion, and the latter was handling specifically schemas and only recursing with itself. The fix was to properly recurse back to `toUserType`, but it required some dodgy stuff surrounding other usage sites like at line 595, which is the reason for the weird Either type. It's not a totally satisfying resolution but it fixes this issue for now.